lib/puppet/parser/functions/pw_hash.rb

```diff
@@ -0,0 +1,43 @@
+Puppet::Parser::Functions::newfunction(
+  :pw_hash,
+  :type => :rvalue,
+  :arity => 3,
+  :doc => "Hashes a password using the crypt function. Provides a hash
+  usable on most POSIX systems.
+
+  The first argument to this function is the password to hash. If it is
+  undef or an empty string, this function returns undef.
+
+  The second argument to this function is a specifier of which hash type
+  to use. Valid hash types are:
+
+  |Specifier|Hash type            |
+  |---------|---------------------|
+  |1        |MD5                  |
+  |5        |SHA-256              |
+  |6        |SHA-512 (recommended)|
+
+  The third argument to this function is the salt to use.
+
+  Note: this uses the Puppet Master's implementation of crypt(3). If your
+  environment contains several different operating systems, ensure that they
+  are compatible before using this function.") do |args|
+    raise ArgumentError, "pw_hash(): wrong number of arguments (#{args.size} for 3)" if args.size != 3
+    raise ArgumentError, "pw_hash(): first argument must be a string" unless args[0].is_a? String
+    raise ArgumentError, "pw_hash(): #{args[1]} is not a valid hash type" unless ['1', '5', '6'].include? args[1].to_s
+    raise ArgumentError, "pw_hash(): third argument must be a string" unless args[2].is_a? String
+    raise ArgumentError, "pw_hash(): third argument must not be empty" if args[2].empty?
+    raise ArgumentError, "pw_hash(): characters in salt must be in the set [a-zA-Z0-9./]" unless args[2].match(/\A[a-zA-Z0-9.\/]+\z/)
+
+    password = args[0].to_s
+    return nil if password.empty?
+
+    # work around JRuby bug in String#crypt for JRuby < 1.7.17
+    if RUBY_PLATFORM == 'java' and 'test'.crypt('$1$1') != '$1$1$Bp8CU9Oujr9SSEw53WV6G.'
```

Comment:
We salt everywhere.  See [line 42](https://github.com/puppetlabs/puppetlabs-stdlib/pull/408/files#diff-dfa59737972d9adbac64e202167f383eR42).  This just checks for a faulty `String#crypt` implementation that doesn't support extended hashes, which is known to exist in older versions of JRuby, and overrides it with a proper Java implementation of `crypt()`.  This weak implementation also exists on MRI on Windows, but we don't have a proper Java implementation under that scenario.  I could change this to fail if the implementation is weak and we're not on Java, though.  That seems like it would probably be a good idea.